Revamp has_request_variables and validators

PIG208 opened the pull request:
<!-- Describe your pull request here.-->
This PR tracks the progress of the proposal on [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/New.20has_request_variables.20API) to revamp the `REQ` family.

TODO:

- [x] Refine error messages from validation failures
- [x] Implement error handling with a wrapper around `cattrs` errors
- [x] Implement the view function decorator
- [x] Migrate `zerver.views.message_flags` and `zerver.views.realm_playgrounds` as an experiment
- [ ] Documentation
- [x] Support `ignored_parameters_unsupported`
- [x] Test cases for the decorator itself
- [ ] Migrate more views and validators

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PIG208 commented:
I'm not sure if we want to support aliases in this prototype, mostly because it is not widely used. I will leave it for now.

PIG208 commented:
Added some examples of usage of this decorator.
We still need to think of a name for it. `endpoint` or maybe something else. I also want to get a better name than `Metadata`.

PIG208 commented:
I added a substantial amount of converter test cases and implemented error handling. It should cover most failure situations now by responding to the client with a translated error message.

showell commented on zerver/lib/converters.py:
```diff
@@ -0,0 +1,269 @@
+    return value
+
+
+def structure_general(value: object, typ: Type[T]) -> T:
```
I don't have huge objections to the use of "typ" as a variable, since it does have the advantage of brevity and it's pretty easy to understand from context.  But I would consider using things like "generic_type" or "primitive_type" depending on the context.

PIG208 replied:
Yeah, `target_type` would be better.

showell commented on zerver/actions/realm_playgrounds.py:
```diff
@@ -38,10 +38,10 @@ def check_add_realm_playground(
-    # check_pygments_language, etc are included in full_clean.
-    # Because we want to avoid raising ValidationError from this check_*
-    # function, we do error handling here to turn it into a JsonableError.
+    # The additional validations using url_template_validaton etc are included
```
etc -> etc.

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+    metadata: Metadata
+    # Name of the function parameter as defined in the original function.
+    param_name: str
+    # Inspected underlying type of the parameter unwarping the Annotated type if
```
unwarp -> unwrap

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+    """This is responsible for inspecting the function signature and getting the
+    metadata from the parameters. We want to keep this function as pure as
+    possible not leaking side effects to the global state. Side effects should
+    be emitted separately after the ViewFuncInfo is returned.
```
emitted -> executed? (The word "emit" feels a little funny here, but I could be wrong. Feel free to disregard.)

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+        # with Optional. We explicitly unwrap it for the case of Annotated,
+        # which otherwise causes undesired behaviors.
+        # This is fixed in Python 3.11: https://github.com/python/cpython/issues/90353
+        param_type = type_hints[param_name]
```
My personal preference would be to extract this chunk of code into an inner helper function called `process_single_parameter` or something.  There are tradeoffs here, so this is a suggestion not a request.

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+            # For path_only parameters, they should already have been passed via
+            # the URL, so there's no need for us to do anything.
+            #
+            # TODO: Run validators for path_only parameters for NewType.
```
Is this TODO for later or is it something you need to address before we merge?

PIG208 replied:
The TODO is the same one from request.py. We leave it for later.

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+            # the request.
+            # This is covered by test_json, but coverage.py
+            # fails to recognize this for some reason.
+            continue  # nocoverage
```
Python does optimizations with `continue` that confuse `coverage.py`.  Whenever there's a non-contorted way to avoid `continue` in Python, which usually entails extracting some kind of helper function or re-arranging conditionals, I try to go that route.  If that's impractical here, then I think this comment is sufficient.  But I'd give it a quick look to see if there's a better way to phrase this code.  Long if/elif/elif/elif blocks are also kind of a code smell, but, again, sometimes unavoidable.

PIG208 replied:
Added a `fallback_to_defaults` dict before this line to make coverage work.

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+            func_param.metadata.documentation_status is DocumentationStatus.DOCUMENTED
+            and not func_param.metadata.path_only
+        ):
+            # TODO: Move arguments_map to here once zerver.lib.request does not
```
Same question about "TODO" as before.

PIG208 replied:
This is also the same TODO item from `has_request_variables`. We plan to work on that separately.

showell commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,334 @@
+                list(request.POST.keys()) + list(request.GET.keys())
+            ).difference(request_notes.processed_parameters)
+
+            # This will be called each time a function decorated with endpoint
```
endpoint -> `@endpoint`

showell commented on zerver/lib/converters.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def format_exception(exc: BaseException, target_type: Union[Type[object], None]) -> str:
+    """The Zulip exception formatter, handling the most common exceptions
```
Can you change the tenses here a bit?  "This is the Zulip exception formatter, which handles the most common exceptions.  It turns exceptions into a translated error message if it recognizes them; otherwise, it re-raises them...."

showell commented on zerver/lib/converters.py:
```diff
@@ -0,0 +1,269 @@
+
+# Converter extensions
+
+# The values in this dict are factory functions that each construcst a hook
```
construcst -> construct

showell commented on zerver/tests/test_converters.py:
```diff
@@ -0,0 +1,226 @@
+        target_type: Type[object],
+        expected_messages: Sequence[str] = [],
+        use_json: bool = True,
+    ) -> object:
```
I really try to avoid boolean arguments in test functions.  Would it be easier to just have two helper functions here?  I know you'd duplicate some of the exception handling code, but is the exception handling code really necessary for somebody to debug the test?

PIG208 replied:
This is used internally by the `verify_formatter_output` helper, and the exception handling here is essential to capture the validation errors raised by `cattrs`.

showell commented on zerver/tests/test_converters.py:
```diff
@@ -0,0 +1,226 @@
+        errors: Sequence[str] = [],
+        output: object = NoOutput,
+    ) -> None:
+        if converter in ("json", "both"):
```
This is a another example of where I would try to avoid conditionals in tests.  You basically could have `verify_json_formatter_output` and `verify_regular_formatter_output`, and if you need to verify both, just call both.

PIG208 replied:
Yeah, I split it into `verify_formatter_output_json` and `verify_formatter_output_regular`.

showell commented:
Is it possible that some code fixes got reverted here?  I see some resolved issues where the typo is still there, despite the fact that there have been some re-pushes since my review.

PIG208 commented:
Hmm, which comments are you referring to? I double-checked and all of the resolved ones are fixed.

PIG208 commented:
Oh I see, some of the converters fixes are pushed to the wrong commit. Let me fix that.

showell commented:
> Hmm, which comments are you referring to? I double-checked and all of the resolved ones are fixed.

Never mind, I was confused by the GitHub UI.  Sorry for the false alarm!

showell commented on zerver/lib/converters.py:
```diff
@@ -162,6 +242,40 @@ def configure_converter(converter: BaseConverter) -> None:
+
+
+# To be explicit about what unions are allowed to be handled by the generic
+# structure_union hook, we whitelist them here.
```
"we whitelist them here" -> "we specify the exact concrete union types that are necessary in Zulip. (Zulip typically has union types related to backward compatibility of APIs that originally took strings but now take ids.  We generally want to phase out union types where possible.)"

(Some folks object to the terms "whitelist" and "blacklist".  While I don't necessarily agree with their objections, you can usually use more precise language like allow-list and deny-list, or, as in this case, just use a more detailed explanation.)

PIG208 replied:
Thanks. I moved `allowed_unions` to the top of the file and added the explanation there.

andersk commented:
To keep this discussion updated—Pydantic 2 was released a month ago, completely rewritten in Rust with 5–50× faster performance and a revamped API that seems to already include many of the features we’re having to build ourselves for cattrs. @PIG208 and I are investigating whether it might be a better alternative now.

Fatal1ty commented:
> To keep this discussion updated—Pydantic 2 was released a month ago, completely rewritten in Rust with 5–50× faster performance and a revamped API that seems to already include many of the features we’re having to build ourselves for cattrs. @PIG208 and I are investigating whether it might be a better alternative now.

I’m interested to see what features you needed in cattrs and check if they exist in [mashumaro](https://github.com/Fatal1ty/mashumaro#benchmark). But if Zulip operate with user data and relies on strict validation rather than coercion, then pydantic 2 will be better choice since both cattrs and mashumaro do coercion by default.

timabbott commented on zerver/tests/test_decorators.py:
```diff
@@ -2343,3 +2363,373 @@ def view_A(
+ParamTypes = Literal["none", "json_only", "both"]
+
+
+class TestEndpoint(ZulipTestCase):
```
This test file is kinda huge. It might be a helpful prep PR to split `test_decorators.py` up. I think we could do:
* `test_validators.py` for `ValidatorTestCase` to parallel `zerver/lib/validator.py`
* And then tests might go into a new `test_api_parameter_parsing.py` (and maybe your new `api.py` would be better named `zerver/lib/api_parameter_parsing.py` or something?

(There's other things we might be able to split out, like a `test_auth_decorators.py`, but I'm not certain of what a good division might be and that may be more clear after this project)

PIG208 replied:
Yeah, the name `api.py` is tentative.

PIG208 replied:
I think some test cases are included in `test_decorators` not because they test decorators, but because they are used in one of the authentication decorators. It is pretty unclear.

timabbott commented on zerver/lib/api.py:
```diff
@@ -0,0 +1,389 @@
+    documentation_status: DocumentationStatus = DocumentationStatus.DOCUMENTED
+    # This can be used to override Pydantic's configuration to use strict mode.
+    # For endpoint parameters wrapped in pydantic.Json, we default it to True,
+    # otherwise it is default to None to use Pydantic's default.
```
This comment might benefit from a few words about what effect it has, or if that's too complex, a link to the documentation, just to make it easy to follow.

(Probably ultimately we'll want to document all the flags here concisely and clearly, since it's a bit of the API)

andersk commented on zerver/lib/api_parameter_parsing.py:
```diff
@@ -0,0 +1,392 @@
+        for error in exc.errors():
+            error_template = ERROR_TEMPLATES.get(error["type"])
+            var_name = parameter.request_var_name + "".join(
+                [f'["{loc}"]' if isinstance(loc, str) else f"[{loc}]" for loc in error["loc"]]
```
Just `f"[{loc!r}]"` would handle numbers and strings with correct escaping.

PIG208 replied:
Right, thanks!

PIG208 commented:
Added parameterized test cases for every known error type. We can use this to verify that each error type can be reached in a certain way.

PIG208 commented:
I modified the way parameterized test cases intended to work ([previously all error types were covered](https://github.com/zulip/zulip/commit/42af4f699b986e60359c358aaae69bc96278898f~#diff-73ec7c7b335d5a9fcb0473ae8342754314933d1efa42bf41a3badc754020c8f2R407-R601)) to avoid unnecessarily testing Pydantic.

PIG208 commented:
I migrated all webhooks to use `@endpoint`.

gnprice commented:
Neat!

A nit: I'd suggest keeping the last commit separate:

    fixup! webhooks: Migrate all webhooks to use endpoint.
    
    This contains all the interesting refactorings that are more than type
    annotation changes of the payload.

rather than squashing it into the big commit with the 77 other more-boring migrations. That's helpful to the reader, because that way the interesting ones are easier to see and don't get hidden amongst the boring ones. They're all independent of each other and don't need to happen atomically, right?

(Personally in fact I'd go further and have a separate commit for each of the interesting migrations, or each of the groups of them for the different ways they can be interesting — the different things one would say about them in the commit message.)

PIG208 commented:
Thanks for the review! I split the commits for webhooks separately with more explanations.

PIG208 commented:
Some updates to `api: Add new endpoint decorators.`:
- Added support for request variable aliases
- Added more doc-strings

PIG208 commented:
Renamed the module to `api_common` with `filter-branch`.

PIG208 commented:
TODO for tommorrow:
- Improve the performance of `parse_param_to_type` with more profiling work
- Rework the aliases implementation

PIG208 commented:
Once the CI passes. It should be ready for review.

PIG208 commented:
No coverage on Python 3.11 because we explicitly check for the version. I guess it is fine to remove that check.

PIG208 commented:
Hmm, checking version info is necessarry, as the test suite for Python 3.11 otherwise fails.

PIG208 commented:
Pushed to remove the unused `CallInfo` class and adjust references to `endpoint` to `typed_endpoint` in some comments.

PIG208 commented:
Rebased to resolve conflicts.

andersk commented on zerver/lib/validator.py:
```diff
@@ -632,6 +634,25 @@ class WildValue:
+        handler: ModelWrapValidatorHandler["WildValue"],
+        info: ValidationInfo,
+    ) -> "WildValue":
+        # TODO: Move WildValue elsewhere to fix circular import
+        from zerver.lib.typed_endpoint import get_func_param
+
+        var_name = get_func_param(info).request_var_name
```
The `get_func_param` context mechanism adds more complexity than is warranted to compute a parameter name that (a) is only used in error messages that the user will never see, (b) doesn’t even name a real parameter since it’s the request body, and (c) is always `payload`, as far as I can see.

We can just hard-code `payload`.

andersk commented on zerver/lib/validator.py:
```diff
@@ -632,6 +634,25 @@ class WildValue:
+        from zerver.lib.typed_endpoint import get_func_param
+
+        var_name = get_func_param(info).request_var_name
+        if isinstance(value, list):
+            return WildValueList(var_name, value)
+        if isinstance(value, dict):
+            return WildValueDict(var_name, value)
+        return WildValue(var_name, value)
```
This duplicates the `wrap_wild_value` function; we can call that instead.

PIG208 commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,558 @@
+
+    for param_name, parameter in parameters.items():
+        assert param_name in type_hints
+        if param_name in IGNORED_PARAMS:
+            continue
```
We don't support or are planning to support passing arguments positionally. Alternative 1 is more sensible to me.

PIG208 commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,558 @@
+        # metadata attached to Annotated. Note that we do not transform
+        # param_type to its underlying type because the Annotated metadata might
+        # still be needed by other parties like Pydantic.
+        ignored_type, *annotations = get_args(param_type)
+        for annotation in annotations:
+            if type(annotation) is Json:
+                json_wrapper = True
```
Nested `Annotated` is [flattened](https://docs.python.org/3/library/typing.html#:~:text=Nested%20Annotated%20types%20are%20flattened.%20The%20order%20of%20the%20metadata%20elements%20starts%20with%20the%20innermost%20annotation%3A), so we should be able to get away with the for loop.

PIG208 commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,558 @@
+    if param_config is None:
+        param_config = ApiParamConfig()
+
+    # Enable strict mode if the parameter is annotated with the pydantic.Json
+    # wrapper, because we want to avoid type coercion. Note that this is
+    # insufficient to handle the case when the Json wrapper is nested inside a
+    # Python object like List[Json[SomeModel]], which is some insanity that we
+    # want to avoid.
+    strict_mode = param_config.strict
+    if strict_mode is None and json_wrapper:
+        strict_mode = True
```
`NonNegativeInt`, or most of the things that originally require a REQ converter will need to be wrapped in `Json`, which shouldn't be an issue against turning strict mode on all the time.

PIG208 replied:
We should note that no user request will be able to provide a valid value for types other than `str` if strict mode is enabled, unless `Json` or something of the same nature is used.

PIG208 replied:
strict mode is now always enabled.

PIG208 commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,558 @@
+}
+
+
+def parse_param_to_type(parameter: FuncParam[T], value: object) -> T:
```
Right. Maybe `parse_parameter_value`?

PIG208 commented:
TBD: Update webhooks and enable strict mode.

andersk commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,596 @@
+
+
+def get_func_param(info: ValidationInfo) -> FuncParam[object]:
+    """To access the FuncParam[T] object created when we call parse_param_to_type
```
`parse_param_to_type` was renamed.

PIG208 commented:
Updates since [c422e98](https://github.com/zulip/zulip/commit/c422e988208f70bef34a250a0a4df100e6a4b456):
- Added `typed_endpoint_without_parameters`, and reworked parameter ignoring using keyword-only parameters.
- Removed the ability to override strict mode via `ApiParamConfig`, and enabled strict mode all the time regardless of `Json`
- Improved exception handling for Pydantic validation error.
- Made `ApiParamConfig` picklable
- Renamed `parse_param_to_type` to `parse_value_for_parameter`
- Rewrote `to_wild_value`

timabbott commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,514 @@
+            param_config = annotation
+    # Set param_config to a default early to avoid additional None-checks.
+    if param_config is None:
+        param_config = ApiParamConfig()
```
How does this empty `ApiParamConfig` get used? Is there an example I should be thinking about?

PIG208 replied:
This is used for parameters that do not define `ApiParamConfig`. It applies to the majority of parameters typed as `str`, `Json[int]` and similar.

timabbott replied:
OK, that makes sense -- probably worth a nice comment explaining this detail, and maybe also a mention of that detail in the `ApiParamConfig` definition that an empty one is used for parameters with no special configuration beyond their type.

timabbott commented on zerver/lib/typed_endpoint.py:
````diff
@@ -0,0 +1,514 @@
+    ...
+```
+
+It is likely a programming error unless you only need the ignored parameters check.
````
Should this say:

``` quote
This is likely a programming error. See https://peps.python.org/pep-3102/ for details on how
to correctly declare your parameters as keyword-only parameters.
Endpoints that do not accept parameters should use @typed_endpoint_without_parameters.
```

PIG208 replied:
Yeah it's definitely nice to have the link there.

timabbott commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,514 @@
+                elif current_alias in request.GET:
+                    value_to_parse = request.GET[current_alias]
+                else:
+                    # This is covered by test_REQ_aliases, but coverage.py
```
The test name here looks copied from the original `REQ` implementation.

PIG208 replied:
Indeed, should be `test_aliases` from the other file.

PIG208 commented on zerver/lib/typed_endpoint.py:
```diff
@@ -0,0 +1,574 @@
+    assert (
+        info.context is not None and "func_param" in info.context
+    ), "The FuncParam object should always be passed as context when validating"
+    return info.context["func_param"]
```
Removed the context information on the type adapter as well.

PIG208 commented:
*Updates*:
- Removed unused `get_func_param` and the context information of the parameter.
- Reworded the error message for the keyword-only requirement of function parameters.
- Refine the Pydantic model defined for the Travis webhook.

timabbott commented on zerver/tests/test_typed_endpoint.py:
```diff
@@ -0,0 +1,596 @@
+        result = call_endpoint(
+            view3, HostRequestMock({"foo": orjson.dumps(False).decode()}), hamlet
+        )
+        self.assertFalse(result)
```
It might be worth a comment explaining what's supposed to be going on with `view3`; or maybe better would be to rename it if that can clarify usefully.

Might be worth a pass generally; I find `def annotated_with_extra_unrelated_annotations` or whatever to be nicer than comments for making tests like this readable.

timabbott reviewed (commented):
Finished a read through the "api: Add new typed_endpoint decorators."; here's the last few comments from that read.

timabbott commented on tools/semgrep.yml:
```diff
@@ -186,3 +186,26 @@ rules:
+    message: |
+      Annotated types containing zerver.lib.typed_endpoint.ApiParamConfig should not be nested inside Optional. Use Annotated[Optional[...], zerver.lib.typed_endpoint.ApiParamConfig(...)] instead.
+    languages: [python]
+    severity: ERROR
```
Semgrep can be kinda slow; is it easy to check the performance impact of adding the new `semgrep` rules? I think `lint --only=semgrep --verbose-timing` before/after should work for this.

timabbott commented on zerver/tests/test_typed_endpoint.py:
```diff
@@ -415,6 +426,48 @@ def view3(
+            raise AssertionError
+
+        with self.assertRaisesMessage(ApiParamValidationError, "test is too long"):
+            call_endpoint(no_nesting, HostRequestMock({"test": "long"}))
```
For this kind of test, it might be nice to call it again with a shorter value just to show that it works.

timabbott commented on zerver/lib/typed_endpoint.py:
```diff
@@ -171,26 +207,45 @@ def parse_single_parameter(
-        origin = get_origin(param_type)
+    if (
+        sys.version_info < (3, 11) and param_default is None
+    ):  # nocoverage # We lose coverage of this with Python 3.11+ only
```
Hmm, is the `version_info` check correct? I think we want someone developing on Python 3.11 to not use nesting constructions in their code, since it'll break on other major Python versions...

timabbott commented on zerver/views/message_edit.py:
```diff
@@ -112,19 +113,18 @@ def get_message_edit_history(
 PROPAGATE_MODE_VALUES = ["change_later", "change_one", "change_all"]
```
This value is now unused; not sure what the best option is for cleaning that up; I feel like ultimately we will want that set of options to be a type and the `zerver/actions/message_edit.py` code to stop doing string comparisons on it.

timabbott commented on zerver/lib/drafts.py:
```diff
@@ -1,60 +1,47 @@
+    content = normalize_body(draft_dict.content)
 
-    timestamp = draft_dict.get("timestamp", time.time())
+    timestamp = draft_dict.timestamp or time.time()
```
This migration may is incorrect in the case that an integer/float timestamp of `0` is passed by the client. Might not matter in this case but we should avoid the pattern.

(I try to avoid using falsy constructions in Python wherever possible in production code because it often has implicit bugs like this)

timabbott commented:
This is great; aside from the comments noted above, this lgtm. I feel like the system is significantly more readable than the only `has_request_variables` one, in large part from being more concise.

We will need to do a round of updates to documentation after merging this -- both our docs for how to write a webhook integration and the "developer tutorials" pages, and then do some marketing in #backend for the development community (I think for now we can do a post and say we're working on new documentation but the system largely works the same and you should be able to crib off existing endpoints as needed). 

I don't have any major concerns about this migration and think we should plan to merge this PR once the above little comments are addressed.

PIG208 commented:
Thanks for the reviews! The comments are addressed.

timabbott commented:
Merged, huge thanks for all the work on rewriting this core part of the Zulip web framework!
